powerpc/AsmToJSON.ml

```diff
@@ -0,0 +1,361 @@
+  | Psubfc (ir1,ir2,ir3) -> fprintf oc "{\"Instruction Name\":\"Psubfc\",\"Args\":[%a,%a,%a]}" p_ireg ir1 p_ireg ir2 p_ireg ir3
+  | Psubfe (ir1,ir2,ir3) -> fprintf oc "{\"Instruction Name\":\"Psubfe\",\"Args\":[%a,%a,%a]}" p_ireg ir1 p_ireg ir2 p_ireg ir3
+  | Psubfze (ir1,ir2) -> fprintf oc "{\"Instruction Name\":\"Psubfe\",\"Args\":[%a,%a]}" p_ireg ir1 p_ireg ir2
+  | Psubfic (ir1,ir2,c) -> fprintf oc "{\"Instruction Name\":\"Psubfic\",\"Args\":[%a,%a,%a]}" p_ireg ir1 p_ireg ir2 p_constant c
+  | Psync -> fprintf oc "{\"Instruction Name\":\"Psync\",\"Args\":[]}"
+  | Ptrap -> fprintf oc "{\"Instruction Name\":\"Ptrap\",\"Args\":[]}"
+  | Pxor (ir1,ir2,ir3) -> fprintf oc "{\"Instruction Name\":\"Pxor\",\"Args\":[%a,%a,%a]}" p_ireg ir1 p_ireg ir2 p_ireg ir3
+  | Pxori (ir1,ir2,c) -> fprintf oc "{\"Instruction Name\":\"Pxori\",\"Args\":[%a,%a,%a]}" p_ireg ir1 p_ireg ir2 p_constant c
+  | Pxoris (ir1,ir2,c) -> fprintf oc "{\"Instruction Name\":\"Pxoris\",\"Args\":[%a,%a,%a]}" p_ireg ir1 p_ireg ir2 p_constant c
+  | Plabel l -> fprintf oc "{\"Instruction Name\":\"Plabel\",\"Args\":[%a]}" p_label l
+  | Pbuiltin (ef,args1,args2) ->
+      begin match ef with
+      | EF_inline_asm (i,s,il) ->
+          fprintf oc "{\"Instruction Name\":\"Inline_asm\",\"Args\":[%a%a%a%a]}" p_atom_constant i  (p_list_cont p_char_list) il
+            (p_list_cont p_preg) args1  (p_list_cont p_preg) args2
+      | _ -> (* Should all be folded away *)
+          assert false
+      end
+  | Pannot _  (* We do not check the annotations *)
+  | Pcfi_adjust _  (* Only debug relevant *)
+  | Pcfi_rel_offset _ ->  () (* Only debug relevant *)
+
+let p_storage oc sto =
+  let storage = match sto with
+  | C.Storage_default -> "Default"
+  | C.Storage_extern -> "Extern"
+  | C.Storage_static -> "Static"
+  | C.Storage_register -> "" (* This should not occur *) in
+  fprintf oc "\"%s\"" storage
+
+let p_section oc = function
+  | Section_text -> fprintf oc "{\"Section Name\":\"Text\"}"
+  | Section_data init -> fprintf oc "{\"Section Name\":\"Data\",\"Init\":%B}" init
+  | Section_small_data init -> fprintf oc "{\"Section Name\":\"Small Data\",\"Init\":%B}" init
+  | Section_const init -> fprintf oc "{\"Section Name\":\"Const\",\"Init\":%B}" init
+  | Section_small_const init -> fprintf oc "{\"Section Name\":\"Small Const\",\"Init\":%B}" init
+  | Section_string -> fprintf oc "{\"Section Name\":\"String\"}"
+  | Section_literal -> fprintf oc "{\"Section Name\":\"Literal\"}"
+  | Section_jumptable -> fprintf oc "{\"Section Name\":\"Jumptable\"}"
+  | Section_user (s,w,e) -> fprintf oc "{\"Section Name\":%s,\"Writable\":%B,\"Executable\":%B}" s w e
+  | Section_debug_info 
+  | Section_debug_abbrev -> () (* There should be no info in the debug sections *)
+
+let p_int_opt oc = function
+  | None -> fprintf oc "0"
+  | Some i -> fprintf oc "%d" i
+
+
+let p_fundef oc (name,f) = 
+  let info = Hashtbl.find C2C.decl_atom name in
```

Comment:
It is very dangerous to call Hashtbl.find without catching the Not_found exception.  To recover the sections associated with id, better call C2C.atom_sections, which is correctly hardened.